libraries/botbuilder-repo-utils/src/workspace.ts

```diff
@@ -3,31 +3,146 @@
+                if (relPath[0] === path.sep) {
+                    relPath = relPath.slice(1);
+                }
+
+                if (filters.path && !minimatch(relPath, filters.path)) {
+                    return;
+                }
+
+                if (filters.ignorePath && minimatch(relPath, filters.ignorePath)) {
+                    return;
+                }
+
                 const pkg = await readJsonFile<Package>(absPath);
                 if (!pkg) {
                     return undefined;
                 }
 
-                return { absPath, pkg };
+                if (filters.noPrivate && pkg.private) {
+                    return;
+                }
+
+                if (filters.script && !pkg.scripts[filters.script]) {
+                    return;
+                }
+
+                if (filters.name && !minimatch(pkg.name, filters.name)) {
+                    return;
+                }
+
+                if (filters.ignoreName && minimatch(pkg.name, filters.ignoreName)) {
+                    return;
+                }
+
+                return { absPath, pkg, relPath };
             }
         )
     );
 
     return R.compact(maybeWorkspaces);
 }
+
+// Takes a set of workspaces and exposes an execute method that invokes a callback
+// concurrently on each workspace, ensuring that dependencies are processed before
+// their consumers.
+export class DependencyResolver {
+    private readonly byName: Record<string, Workspace>;
+    private readonly depsByName: Record<string, Workspace[]>;
+
+    constructor(private readonly workspaces: Workspace[], private readonly concurrency = Infinity) {
+        this.byName = this.workspaces.reduce<Record<string, Workspace>>(
+            (acc, workspace) => ({ ...acc, [workspace.pkg.name]: workspace }),
+            {}
+        );
+
+        this.depsByName = this.workspaces.reduce<Record<string, Workspace[]>>((acc, workspace) => {
+            const deps = [
+                ...Object.keys(workspace.pkg.dependencies ?? {}),
+                ...Object.keys(workspace.pkg.devDependencies ?? {}),
+            ]
+                .map((name) => this.byName[name])
+                .filter((pkg) => !!pkg);
+
+            return { ...acc, [workspace.pkg.name]: deps };
+        }, {});
+    }
+
+    async execute(callback: (workspace: Workspace, group: number) => Promise<unknown>): Promise<void> {
+        const resolved = new Set<string>();
+
+        // Returns list of packages where all dependencies are resolved
+        const ready = () =>
+            R.flatten(
+                Object.entries(this.byName)
+                    .filter(([name]) => {
+                        const deps = this.depsByName[name];
+
+                        const needs = deps.filter((dep) => !resolved.has(dep.pkg.name));
+
+                        // Only report a cycle if both packages are just waiting for each other
+                        const cycle =
```

Comment:
We actually do have a cycle, `azure` (`tools/package.json`) references `botframework-connector` which, itself, references `azure` which is why this code exists.